File: java/longest_valid_parentheses.java
public class Solution {
    public int longestValidParentheses(String s) {
        int longest=0;
        int left=-1;
        Stack<Integer> stack=new Stack<>();
        for (int i=0; i<s.length(); i++) {
            if (s.charAt(i)=='(') stack.push(i);
            else {
                if (stack.isEmpty()) left=i;
                else {
                    stack.pop();
                    if (stack.isEmpty()) longest=Math.max(longest, i-left);
                    else longest=Math.max(longest, i-stack.peek());
                }
            }
        }
        return longest;
    }
}

File: python/longest-valid-parentheses.py
class Solution:
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        stk, res=[], 0
        for i in range(len(s)):
            if s[i]=='(': stk.append(i)
            else:
                if len(stk)==0 or s[stk[-1]]==')':
                    stk.append(i)
                else:
                    stk.pop()
                    res=max(res, (i+1 if len(stk)==0 else i-stk[-1]))
        
        return res
